### PR TITLE
8332670: C1 clone intrinsic needs memory barriers

### PR DESCRIPTION
**Pull request by galderz**
Adds a storestore barrier after copying the contents in the primitive array intrinsic (credit @shipilev). The barrier is a no-op in platforms where not needed so no need for an ifdef.

The barrier after new array creation is only added if zeroing the array on aarch64 (credit @dean-long). Since the primitive array clone intrinsic does not zero the array, that means there's a single barrier added for this use case.

There's no barrier added on x86 c1 macro assembler for nothing to do there. 

I've run the following tests:
* tier 1 on darwin/aarch64
* tier 1 on linux/x86_64
* `hotspot_compiler` tests on darwin/aarch64
* `copy.clone.arrays` jcstress tests on darwin/aarch64.

I tried but was unable to create a standalone test for the jdk source tree that would fail.

FYI @bulasevich @TheRealMDoerr @RealFYang @RealLucy similar platform specific c1 macro assembler changes might be required for other platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332670](https://bugs.openjdk.org/browse/JDK-8332670): C1 clone intrinsic needs memory barriers (**Bug** - P2)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19538/head:pull/19538` \
`$ git checkout pull/19538`

Update a local copy of the PR: \
`$ git checkout pull/19538` \
`$ git pull https://git.openjdk.org/jdk.git pull/19538/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19538`

View PR using the GUI difftool: \
`$ git pr show -t 19538`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19538.diff">https://git.openjdk.org/jdk/pull/19538.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19538#issuecomment-2146897565)

### CODE REVIEW

**Comment by galderz** [comment]
Forgot to say, this change shows no impact on the array clone micro benchmark. This is expected since the change moves the barrier from one place (after array creation) to another (after copying contents), so no additional barriers are introduced.

File: src/hotspot/cpu/aarch64/c1_MacroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -304,7 +304,12 @@ void C1_MacroAssembler::allocate_array(Register obj, Register len, Register t1,
     return;
   }
 
-  membar(StoreStore);
+  // Only add membar if zeroing the array.
+  // If not zeroing, subsequent instructions should populate the array (e.g. copy contents),
+  // and the membar should be set after the array has been populated accordingly.
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c1_MacroAssembler_aarch64.cpp:309]: Strengthen the language:
```suggestion
  // Only add membar if zeroing the array.
  // If not zeroing, subsequent instructions must fully populate the array (e.g. copy contents),
  // and a membar must be emitted after the array has been populated accordingly. See
  // GraphBuilder::append_alloc_array_copy.
```

**Review by shipilev — APPROVED** [review]
All right, that looks reasonable.

I am a bit queasy on removing the storestore barrier from the allocation path, given that it also protects the object metadata. An accidentally missing barrier would probably lead to VM crash, that is in the best case. Current code paths do not seem to be affected by this, but there is also no guardrails that would protect us from making such a mistake in the future: someone adds `new NewTypeArray` somewhere, and forgets a trailing barrier? 

What would be the cost of still emitting (an excess for cloning path) StoreStore in `C1_MacroAssembler::allocate_array`? Would that cost still matter, given the performance improvement we get with C1 clone intrinsic?

**Comment by theRealAph** [comment]
> All right, that looks reasonable.
> 
> I am a bit queasy on conditionally removing the `StoreStore` barrier from the allocation path, given that it also protects the object metadata. An accidentally missing barrier would probably lead to VM crash, that is in the best case. Current code paths do not seem to be affected by this, but there is also no guardrails that would protect us from making such a mistake in the future: someone adds `new NewTypeArray` somewhere, and forgets a trailing barrier?

I agree with you. I suspect an additional StoreStore` barrier after writing the header would make no difference in performance.

**Comment by TheRealMDoerr** [comment]
PPC64 adds storestore barriers in LIR (not in assembler as aarch64): https://github.com/openjdk/jdk/blob/8d3de45f4dfd60dc4e2f210cb0c085fcf6efb8e2/src/hotspot/cpu/ppc/c1_LIRGenerator_ppc.cpp#L928
I think this fits nicely to your change.

**Comment by zifeihan** [comment]
@galderz Hi, I have finished the RISC-V part, tier1-3 tested on SOPHON SG2042.
Please help us to add the RISC-V part, thanks a lot!
[19538-riscv-port-v1.diff.txt](https://github.com/user-attachments/files/15571986/19538-riscv-port-v1.diff.txt)

**Comment by galderz** [comment]
> All right, that looks reasonable.
> 
> I am a bit queasy on conditionally removing the `StoreStore` barrier from the allocation path, given that it also protects the object metadata. An accidentally missing barrier would probably lead to VM crash, that is in the best case. Current code paths do not seem to be affected by this, but there is also no guardrails that would protect us from making such a mistake in the future: someone adds `new NewTypeArray` somewhere, and forgets a trailing barrier?
> 
> What would be the cost of still emitting (an excess for cloning path) `StoreStore` in `C1_MacroAssembler::allocate_array`? Would that cost still matter, given the performance improvement we get with C1 clone intrinsic?

Let me make sure I understand this right: you are suggesting removing the `if (zero_array)` branch and instead always emit storestore barrier, as well as the storestore barrier at the end of `append_alloc_array_copy`?

**Comment by shipilev** [comment]
> Let me make sure I understand this right: you are suggesting removing the `if (zero_array)` branch and instead always emit storestore barrier, as well as the storestore barrier at the end of `append_alloc_array_copy`?

Yes. AFAICS, we would have an excess `StoreStore` only on C1 clone path, which is should be fast enough already after you introduced the intrinsic. That is to say, that even if that `StoreStore` regresses performance a little bit in comparison to current state, it should still be faster than the version that did not do C1 clone intrinsic implemented at all.

**Review by shipilev — APPROVED** [review]
Yup, like this: nice and tidy.

**Comment by galderz** [comment]
> Yup, like this: nice and tidy.

Just like you initially suggested! 😅

**Review by TobiHartmann — APPROVED** [review]
This looks good to me too and passed our testing.

**Review by theRealAph — APPROVED** [review]

**Comment by TheRealMDoerr** [comment]
Seems like the `/integrate` got lost.  I can see it in my email.

**Comment by galderz** [comment]
@TheRealMDoerr I removed it because I had just noticed the test failure in CI, looking into it.

**Comment by TheRealMDoerr** [comment]
> @TheRealMDoerr I removed it because I had just noticed the test failure in CI, looking into it.

The same test has failed in one of my PRs. I don't think it is related and can be ignored.

**Comment by galderz** [comment]
> > @TheRealMDoerr I removed it because I had just noticed the test failure in CI, looking into it.
> 
> The same test has failed in one of my PRs. I don't think it is related. Can be ignored.

Ah ok. I've tried to replicate it on a linux env but didn't have luck. I'll go ahead and try to integrate.

**Pull request closed without merging**